src/NuGet.Clients/NuGet.Console/OutputConsole/ChannelOutputConsole.cs

```diff
@@ -0,0 +1,149 @@
+        private readonly string _channelOutputName;
+        private readonly JoinableTaskFactory _joinableTaskFactory;
+
+        private AsyncLazy<ServiceBrokerClient> _serviceBrokerClient;
+        private PipeWriter _channelPipeWriter;
+        private bool _disposedValue = false;
+
+        public ChannelOutputConsole(IAsyncServiceProvider asyncServiceProvider, Guid channelId, string outputName, JoinableTaskFactory joinableTaskFactory)
+        {
+            if (asyncServiceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(asyncServiceProvider));
+            }
+            _channelGuid = channelId;
+            _channelId = _channelGuid.ToString();
+            _channelOutputName = outputName ?? throw new ArgumentNullException(nameof(outputName));
+            _joinableTaskFactory = joinableTaskFactory ?? throw new ArgumentNullException(nameof(joinableTaskFactory));
+
+            _serviceBrokerClient = new AsyncLazy<ServiceBrokerClient>(async () =>
+            {
+                IBrokeredServiceContainer container = (IBrokeredServiceContainer)await asyncServiceProvider.GetServiceAsync(typeof(SVsBrokeredServiceContainer));
+                Assumes.Present(container);
+                IServiceBroker serviceBroker = container.GetFullAccessServiceBroker();
+                return new ServiceBrokerClient(serviceBroker, _joinableTaskFactory);
+            }, _joinableTaskFactory);
+        }
+
+        public override Task ActivateAsync()
+        {
+            // Nothing to do here in Visual Studio Online Environments mode. We do not create the pane
+            return Task.CompletedTask;
+        }
+
+        public override Task ClearAsync()
+        {
+            return CloseChannelAsync();
+        }
+
+        public override Task WriteAsync(string text)
+        {
+            return WriteToOutputChannelAsync(_channelId, _channelOutputName, text, CancellationToken.None);
+        }
+
+        private async Task WriteToOutputChannelAsync(string channelId, string displayNameResourceId, string content, CancellationToken cancellationToken)
+        {
+            using (await _pipeLock.EnterAsync())
+            {
+                if (_channelPipeWriter == null)
+                {
+                    var pipe = new Pipe();
+
+                    using (var outputChannelStore = await (await _serviceBrokerClient.GetValueAsync()).GetProxyAsync<IOutputChannelStore>(VisualStudioServices.VS2019_4.OutputChannelStore, cancellationToken))
+                    {
+                        if (outputChannelStore.Proxy != null)
+                        {
+                            await outputChannelStore.Proxy.CreateChannelAsync(channelId, displayNameResourceId, pipe.Reader, TextEncoding, cancellationToken);
+                            _channelPipeWriter = pipe.Writer;
+
+                            // write any deferred messages
+                            foreach (var s in _deferredOutputMessages)
```

Comment:
The process of deferred messages happens only when the _channelPipeWriter == null .  Is it intended?

Reply:
Yes, because it would've been written otherwise. 
We only queue messages when the writer is null, and then write whenever we get a valid writer again.